Base/Python/slicer/tests/test_slicer_parameter_pack.py

```diff
@@ -0,0 +1,288 @@
+        class ParameterPack:
+            x: int
+            y: Annotated[int, Default(1)]
+
+        pack = ParameterPack()
+        self.assertEqual(pack.x, 0)
+        self.assertEqual(pack.y, 1)
+
+        pack2 = ParameterPack(2, 3)
+        self.assertEqual(pack2.x, 2)
+        self.assertEqual(pack2.y, 3)
+
+        pack3 = ParameterPack(y=2, x=3)
+        self.assertEqual(pack3.x, 3)
+        self.assertEqual(pack3.y, 2)
+
+        pack4 = ParameterPack(6, y=8)
+        self.assertEqual(pack4.x, 6)
+        self.assertEqual(pack4.y, 8)
+
+        pack5 = ParameterPack(6)
+        self.assertEqual(pack5.x, 6)
+        self.assertEqual(pack5.y, 1)
+
+        pack6 = ParameterPack(y=6)
+        self.assertEqual(pack6.x, 0)
+        self.assertEqual(pack6.y, 6)
+
+        with self.assertRaises(ValueError):
+            pack.x = None
+        with self.assertRaises(ValueError):
+            pack.x = "hi"
+
+    def test_parameter_pack_validation(self):
+        @parameterPack
+        class ParameterPack:
+            x: Annotated[float, WithinRange(0, 10)]
+            y: Annotated[str, Choice(["a", "b"]), Default("b")]
+
+        pack = ParameterPack()
+        self.assertEqual(pack.x, 0)
+        self.assertEqual(pack.y, "b")
+
+        with self.assertRaises(ValueError):
+            ParameterPack(-1, "a")
+        with self.assertRaises(ValueError):
+            ParameterPack(10.001, "a")
+        with self.assertRaises(ValueError):
+            ParameterPack(0, "c")
+        with self.assertRaises(ValueError):
+            ParameterPack(None, "a")
+        with self.assertRaises(ValueError):
+            ParameterPack(2.2, 4)
+
+    def test_parameter_pack_nesting(self):
+        box = BoundingBox()
+        self.assertEqual(box.topLeft, Point())
+        self.assertEqual(box.bottomRight, Point())
+
+        box.topLeft = Point(-2, 3)
+        box.bottomRight.x = 2
+        box.bottomRight.y = -3
+
+        self.assertEqual(box, BoundingBox(Point(-2, 3), Point(2, -3)))
+
+    def test_parameter_pack_custom_init_eq_repr_str(self):
+        @parameterPack
+        class ParameterPack:
+            i: int
+            j: int
+            k: str
+
+            def __init__(self, k):
+                self.i = 1
+                self.j = 4
+                self.k = k
+
+            def __str__(self) -> str:
+                return "mystr"
+
+            def __repr__(self) -> str:
+                return "myrepr"
+
+            def __eq__(self, other) -> bool:
+                return self.j == other.j
+
+        pack = ParameterPack("hey")
+        self.assertEqual(pack.i, 1)
+        self.assertEqual(pack.j, 4)
+        self.assertEqual(pack.k, "hey")
+
+        pack.j = 5
+        pack2 = ParameterPack("words")
+        pack2.i = 6
+        pack2.j = 5
+
+        self.assertEqual(pack, pack2)
+
+        self.assertEqual(str(pack), "mystr")
+        self.assertEqual(repr(pack), "myrepr")
+
+    def test_serialization(self):
```

Comment:
Similarly, might be good to ensure you can't dynamically add a property to the observed pack; `param.pack.z = 123`.

Reply:
Good idea, will add.

Reply:
The observed version does prevent this because it overrides `__setattr__`. I can see an argument that this makes sense because setting a new attribute is a change that can't be observed.

Dynamically adding on the non-observed parameterPack may still be ok to allow because it follows dataclass's lead and is a somewhat more generic structure than the observed side.